### ISSUE TITLE
Idea: Pipeline Mutexes

### ISSUE THREAD

**Issue by wlynch** [tektoncd/pipeline#2828]
This was an idea that @k floated to me awhile back, but I finally got around to making an issue to discuss. What I'm curious about:

- Is this a use case we want to focus on?
- Is this worth make this a built-in feature? (as opposed to a Catalog feature)
- Any other features / alternatives to consider

Details intentionally vague - this is a "should we do this?" issue, not a "how we'll do this" issue.

# Idea

I may want to control how Pipelines run in relation to others and ensure only 1 pipeline for a given selector can run at a time (hence a "mutex"). 
I may want to reject new Pipelines if one if a similar one is already running, or queue it up and just make sure it does not run in parallel. This might be because:

- I have a presubmit pipeline that I want to only run one instance at a time per pull request to reduce costs (e.g. in case someone pushes multiple commits. I only need to run the most recent and can cancel the currently running pipelines).
- My pipeline mutates some external state, and I want to make sure only one thing operates on it at a time.

## Possible solution

Have a mechanism to select conditions to allow Pipeline execution, as well as a strategy for what to do in response.

### Examples

If a new Pipeline is created that was labelled as a pull request, cancel existing runs.
```
selector: repo=foo, type=pullrequest
strategy: cancel
```

Only run 1 pipeline at a time that was labelled as being started by a push to master. (does not guarantee ordering)
```
selector: repo=foo, type=push, ref=master
strategy: queue
```

Deny new pipeline create requests if they match a pipeline currently running.
```
selector: repo=foo, type=push, ref=master
strategy: deny
```

# Alternatives

## Implement as a task

- Cancellation could be handled by having the first step of every pipeline could include something along the lines of 

  ```
  kubectl delete pipelinerun -l foo=bar
  ```

  This would clobber over any other Pipelines with a particular label.

- Queueing could be handled by having a Condition that runs `kubectl get` for running pods, and only proceed if a condition is true. This is difficult since you'd have to get creative in inspecting runtime information of other runs (e.g. are they also in a wait state, or are they running). This also creates container waste since the pipelines would all be running.

- Deny could not be implemented this way.

**Comment by tragiclifestories** [comment]
We've built a queueing system to manage our way around this problem, so a +1 for it being a useful thing to tackle. I don't know whether it should be a core primitive or in the catalog, but in our use case is was necessary at a very early stage, and for deployments specifically it seems to me that having one deployment per app per environment at a time, and ideally in a sensible order, is going to be a very common requirement. So I'm leaning towards 'core'.

**Comment by holly-cummins** [comment]
I'd really appreciate this as well, perhaps with Task granularity rather than pipeline. My use case for this, which is to do with cross-talk between concurrent runs of integration tests/db management. In an integration test scenario, for example, the tasks depend on an external resource. If that resource is stateful (like a database), some tasks are rebuilding the database while others might be executing tests which use the database. I'd love to be able to single-thread pipeline runs through the integration test phase.

**Comment by resamaraschi** [comment]
I also think this is a very common use case for a CI/CD Pipeline. 
Our scenario is that we have one test cluster for all the created PRs. For instance when 2 developers opens 2 PRs, the pipeline should test one PR against the test cluster first and set the status for corresponding PR. Meanwhile all other pipelineRuns for other PRs should be queued until the cluster is free for the next run.

**Comment by holly-cummins** [comment]
I got inspired by @tragiclifestories 's suggestion of a queueing system as a workaround, so I made one too. I documented the steps - hopefully it's useful to someone else while this is pending: https://medium.com/@holly.k.cummins/using-lease-resources-to-manage-concurrency-in-tekton-builds-344ba84df297

**Comment by tragiclifestories** [comment]
Interesting! 

We took a different approach by storing the queue data in configmaps and defining all the queue operations as scripts that run in task steps. So no explicit modelling through CRDs but it works well enough for our use case. 

Hopefully we'll get around to the blog-post stage of the project soon.

**Comment by afrittoli** [comment]
> I got inspired by @tragiclifestories 's suggestion of a queueing system as a workaround, so I made one too. I documented the steps - hopefully it's useful to someone else while this is pending: https://medium.com/@holly.k.cummins/using-lease-resources-to-manage-concurrency-in-tekton-builds-344ba84df297

Nice :) 
@pritidesai finally in action

**Comment by PoliM** [comment]
Here is the use case we currently have. Imagine this simplified CD pipeline:
`--> DeployToDev --> TestOnDev --> DeployToPreProd --> TestOnPreProd --> DeployToProd --> SmoketestOnProd`
There are something like three sections Dev, PreProd and Prod. 
The pipeline is started for a commit in a Git repository that holds the configuration of an application (GitOps). Now here are some requirements:
- if a PipelineRun is in `TestOnPreProd` we don't want another PipelineRun to start `DeployToPreProd` because the test should be the result of the configuration that the first PipelineRun was started with.
- but it is ok to have another PipelineRun starting to `DeployToDev`.
- a PipelineRun must not take over another PipelineRun because the CD pipeline should deliver the stuff in the order as the configurations were committed to the Git repository.
- to make things worse - we use the same pipeline to deploy different applications. So the exclusivity should be per section and application.

Currently we use a task in front of the section that polls a REST service to "ask to enter the section". The implementation of the REST service is specific to our pipeline and uses the Tekton API to analyse the state of all the PipelineRuns. It's ugly :blush: but it works so far.

**Comment by Letty5411** [comment]
I'd prefer both task and pipeline granularity mutex.

**Comment by Letty5411** [comment]
Hi @pritidesai , is there any update about this issue? Thanks :)

**Comment by julweber** [comment]
+1

**Comment by julweber** [comment]
Dear tekton Team,

is there any update about this?

It would be great if there was an option to set pipeline runs to a serial mode.
In the scenario that a new pipeline run is added while a pipeline run for the same pipeline within the same namespace is already running, the new run could wait for the existing one to finish before being executed.

For example:
Concourse CI allows this via a serial job - https://concourse-ci.org/serial-job-example.html

Kind regards

**Comment by jerop** [comment]
> is there any update about this?

@julweber this is being explored in tekton experimental repo: https://github.com/tektoncd/experimental/issues/699 - @ImJasonH shared an idea in that issue

**Comment by vdemeester** [comment]
Closing as duplicate of #2591 (Controlling max parallel jobs per pipeline), which is the most comprehensive discussion on concurrency/synchronization/mutex support in Tekton Pipelines.

The core need described here — mutual exclusion and locking for pipeline resources — is covered by the broader concurrency control discussion in #2591. Consolidating there to keep the conversation in one place.
